src/infill/ZigzagConnectorProcessor.cpp

```diff
@@ -0,0 +1,156 @@
+        //   - if direction < 0 (top zags: right -> left), skip zags whose <end> mod 3 == 0
+        //
+        if (direction > 0)
+        {
+            should_skip_this_connection = start_scanline_idx % this->zag_skip_count == 0;
+        }
+        else
+        {
+            should_skip_this_connection = (start_scanline_idx - 1) % this->zag_skip_count == 0;
+        }
+    }
+
+    const bool should_add =
+        (is_this_connection_even && !is_this_endpiece && !should_skip_this_connection) // normal connections that should be added
+        || (this->use_endpieces && is_this_endpiece);  // end piece if it is enabled;
+
+    return should_add;
+}
+
+
+void ZigzagConnectorProcessor::registerScanlineSegmentIntersection(const Point& intersection, int scanline_index, int direction)
+{
+
+    if (this->is_first_connector)
+    {
+        // process as the first connector if we haven't found one yet
+        // this will be processed with the last remaining piece at the end (when the polygon finishes)
+        this->first_connector.push_back(intersection);
+        this->first_connector_end_scanline_index = scanline_index;
+        this->is_first_connector = false;
+    }
+    else
+    {
+        // add this connector if needed
+        if (this->shouldAddCurrentConnector(this->last_connector_index, scanline_index, direction))
+        {
+            const bool is_this_endpiece = scanline_index == this->last_connector_index;
+            this->current_connector.push_back(intersection);
+            this->addZagConnector(this->current_connector, is_this_endpiece);
+        }
+    }
+
+    // update state
+    this->current_connector.clear(); // we're starting a new (odd) zigzag connector, so clear the old one
+    this->current_connector.push_back(intersection);
+    this->last_connector_index = scanline_index;
+}
+
+
+void ZigzagConnectorProcessor::registerPolyFinished()
+{
+    int scanline_start_index = this->last_connector_index;
+    int scanline_end_index = this->first_connector_end_scanline_index;
+    int direction = scanline_end_index > scanline_start_index ? 1 : -1;
+    const bool is_endpiece = this->is_first_connector || (!this->is_first_connector && scanline_start_index == scanline_end_index);
+
+    // decides whether to add this zag according to the following rules
+    if ((is_endpiece && this->use_endpieces)
+        || (!is_endpiece && this->shouldAddCurrentConnector(scanline_start_index, scanline_end_index, direction)))
+    {
+        // for convenience, put every point in one vector
+        for (const Point& point : this->first_connector)
+        {
+            this->current_connector.push_back(point);
+        }
+        this->first_connector.clear();
+
+        this->addZagConnector(this->current_connector, is_endpiece);
+    }
+
+    // reset member variables
+    this->reset();
+}
+
+
+void ZigzagConnectorProcessor::addZagConnector(const std::vector<Point>& points, bool is_endpiece)
```

Comment:
This function is always called with the same first argument, namely `this->current_connector`, so you could leave out the argument from the function signature.
If you do so you could rename the function to `addCurrentZagConnector`